qiskit/pulse/library/symbolic_pulses.py

```diff
@@ -0,0 +1,810 @@
+    t: sym.Symbol,
+    center: Union[sym.Symbol, sym.Expr, complex],
+    t_zero: Union[sym.Symbol, sym.Expr, complex],
+    sigma: Union[sym.Symbol, sym.Expr, complex],
+) -> sym.Expr:
+    r"""Helper function that returns a lifted Gaussian symbolic equation.
+
+    For :math:`\sigma=` ``sigma`` the symbolic equation will be
+
+    .. math::
+
+        f(x) = \exp\left(-\frac12 \left(\frac{x - \mu}{\sigma}\right)^2 \right),
+
+    with the center :math:`\mu=` ``duration/2``.
+    Then, each output sample :math:`y` is modified according to:
+
+    .. math::
+
+        y \mapsto \frac{y-y^*}{1.0-y^*},
+
+    where :math:`y^*` is the value of the un-normalized Gaussian at the endpoints of the pulse.
+    This sets the endpoints to :math:`0` while preserving the amplitude at the center,
+    i.e. :math:`y` is set to :math:`1.0`.
+
+    Args:
+        t: Symbol object representing time.
+        center: Symbol or expression representing the middle point of the samples.
+        t_zero: The value of t at which the pulse is lowered to 0.
+        sigma: Symbol or expression representing Gaussian sigma.
+
+    Returns:
+        Symbolic equation.
+    """
+    gauss = sym.exp(-(((t - center) / sigma) ** 2) / 2)
+    offset = sym.exp(-(((t_zero - center) / sigma) ** 2) / 2)
+
+    return (gauss - offset) / (1 - offset)
+
+
+@functools.lru_cache(maxsize=None)
+def _is_amplitude_valid(symbolic_pulse: "SymbolicPulse") -> bool:
+    """A helper function to validate maximum amplitude limit.
+
+    Result is cached for better performance.
+
+    Args:
+        symbolic_pulse: A pulse to validate.
+
+    Returns:
+        Return True if any sample point exceeds 1.0 in absolute value.
+    """
+    return np.all(np.abs(symbolic_pulse.get_waveform().samples) < 1.0)
+
+
+def _get_expression_args(expr: sym.Expr, params: Dict[str, float]) -> List[float]:
+    """A helper function to get argument to evaluate expression.
+
+    Args:
+        expr: Symbolic expression to evaluate.
+        params: Dictionary of parameter, which is a superset of expression arguments.
+
+    Returns:
+        Arguments passed to the lambdified expression.
+    """
+    args = []
+    for symbol in sorted(expr.free_symbols, key=lambda s: s.name):
+        if symbol.name == "t":
+            # 't' is a special parameter to represent time vector.
+            # This should be place at first to broadcast other parameters
+            # in symengine lambdify function.
+            times = np.arange(0, params["duration"]) + 1 / 2
+            args.insert(0, times)
+            continue
+        args.append(params[symbol.name])
+    return args
+
+
+class LambdifiedExpression:
+    """Descriptor to lambdify symbolic expression with cache.
+
+    When new symbolic expression is set for the first time,
+    this will internally lambdify the expressions and store the callbacks in the instance cache.
+    For the next time it will just return the cached callbacks for speed up.
```

Comment:
```suggestion
    When a new symbolic expression is created for the first time, ``LambdifiedExpression``
    will internally lambdify the expressions and store the resulting callbacks in its cache.
    The next time it encounters the same expression it will return the cached callbacks
    thereby increasing the code's speed.
```
Is `created` the right replacement for `set` ? I could not quite understand what is meant by `set`.

Reply:
It is already created. This class is a descriptor having `__get__` and `__set__` method. So this is literary `set` but might be confusing without knowing how the code works. I replaced it with `assign` and added link to descriptor by following the suggestion from @wshanks